Convert install-annex.sh to Python

jwodder opened the pull request:
Step 1 of #5078

yarikoptic commented on datalad/install.py:
```diff
@@ -0,0 +1,396 @@
+    def _install_via_autobuild_or_snapshot_linux(self, subpath):
+        tmpdir = tempfile.mkdtemp(prefix="ga-")
+        self.annex_bin = os.path.join(tmpdir, "git-annex.linux")
+        print("I: downloading and extracting under {self.annex_bin}", file=sys.stderr)
```
Let's use logging, request a logger for `datalad.install` and if there is no handlers in that logger's "root"  (thus not imported from datalad) it would establish the simple default one (probably dumping to stderr).  This way it would stay nicely consistent with datalad logging when used from "inside" of datalad and not as a script, and would still log nicely if used as a script.

jwodder replied:
Done.

yarikoptic commented on datalad/install.py:
```diff
@@ -0,0 +1,396 @@
+        self._path = newpath
+        if self.env_write_file is not None:
+            with self.env_write_file.open("a") as fp:
+                print(f"PATH={quote(newpath)}", file=fp)
```
shouldn't we  `export PATH` ?

yarikoptic replied:
also why just not to make this `addpath` and thus have here
```suggestion
                print(f"export PATH={quote(newpath)}:$PATH", file=fp)
```
so outside could `source` the file possibly from different environments (thus possibly having different PATH set).  E.g. I already used this installed with snapshots of git-annex within virtualenv and outside etc.

jwodder replied:
Unfortunately, the conda-forge-last schema needs to add its path directory to the *end* of `PATH`.  The comment from the original source reads:

> We are interested only to get git-annex into our environment  
> So as to not interfere with "system wide" Python etc, we will add  
> miniconda at the end of the path

yarikoptic replied:
that is ok, add the option `last=False` and append to `$PATH` instead of prepending `if last`

jwodder replied:
Done.

yarikoptic reviewed (changes requested):
Woohoo!  Thanks! I left some requests. 
But also, since tests would run anyways, in the next batch of pushed commits (datalad CI run takes awhile, so better be discrete with pushes ;)), just adjust all the CI scripts to use this one instead of prior bash version or manual invocation if present. That would immediately test it as well (I guess we might come up with some mocked up or even a few integration tests later on).

jwodder commented:
@yarikoptic I could only find two direct uses of `install-annex.sh` in the repo (one in the `test_macos.yml` workflow, the other in `.travis.yml`).  There's also `tools/ci/prep-travis-devel-annex.sh`, which is almost like the `neurodebian-devel` schema, except it fails if the devel version is not newer than the installed version, but that script isn't even used by anything in the repo.  Are there any other scripts you are aware of that need to be updated to use `install.py`?

yarikoptic commented:
probably only those which you already found ATM.

yarikoptic commented on datalad/install.py:
```diff
@@ -0,0 +1,399 @@
+#!/usr/bin/python3
```
```suggestion
#!/usr/bin/env python3
```
?

jwodder replied:
Done.

yarikoptic commented on datalad/install.py:
```diff
@@ -0,0 +1,399 @@
+        self._install_via_conda(version, tmpdir)
+
+    def _install_via_conda(self, version, tmpdir):
+        miniconda_script = "Miniconda3-latest-Linux-x86_64.sh"
```
note: Linux specific ATM. Later will need to be RFed to support Windows and OSX

yarikoptic reviewed (commented):
shouldn't shebang use `/usr/bin/env`?

jwodder commented:
@yarikoptic Some of the Travis tests are failing because they're using Python 3.5, yet I wrote this script for 3.6.  Python 3.5 reached EOL at the end of September; are you planning to drop support from Datalad, or do I have to give up my precious f-strings?

yarikoptic commented:
oh, sorry -- I missed that. I will change base to become `master` - we dropped 3.5 in there already, and hopefully 0.14.0 is not too far away, so let's look forward and not loose precious `f-strings`

yarikoptic commented:
travis isn't happy due to seems missing `=`:
```
subprocess.CalledProcessError: Command '['/tmp/ga-tsoa94ls/miniconda/bin/conda', 'install', '-q', '-c', 'conda-forge', '-y', 'git-annex7.20190819']' returned non-zero exit status 1.
```
in the last argument between git-annex and its version to install .  Didn't check what appveyor is complaining about.

jwodder commented:
@yarikoptic I fixed the `=` bit.  Appveyor was failing due to running on Python 3.5.  The change of base branch seems to have fixed that, though I also rebased my branch just to be sure.

yarikoptic commented:
Great, thank you @jwodder ! 

@datalad/developers - any objections on having this merged?

Possible TODOs I have spotted for later:
- make it possible to pass `args` to `main`/arg parser  so that `main` could be tested and used within Python
  - add some smoke or unit-tests for the code (possibly mocking out actual calls to installers where possible)
- check if any extensions etc make use of `remove tools/ci/install-annex.sh` in their CI setups, and if not -- remove it to avoid confusion etc

kyleam commented:
> any objections on having this merged?

None from me.
